promql/prettier/node.go

```diff
@@ -0,0 +1,169 @@
+	switch element {
+	case grouping:
+		switch node := n.currentNode.(type) {
+		case *parser.BinaryExpr:
+			if node.VectorMatching == nil {
+				return node.ReturnBool
+			}
+			return len(node.VectorMatching.MatchingLabels) > 0 || node.ReturnBool || node.VectorMatching.On
+		case *parser.AggregateExpr:
+			return len(node.Grouping) > 0
+		}
+	case scalars:
+		if node, ok := n.currentNode.(*parser.BinaryExpr); ok {
+			if node.LHS.Type() == parser.ValueTypeScalar || node.RHS.Type() == parser.ValueTypeScalar {
+				return true
+			}
+		}
+	case multiArguments:
+		if node, ok := n.currentNode.(*parser.Call); ok {
+			return len(node.Args) > 1
+		}
+	}
+	return false
+}
+
+// nodeHistory returns the ancestors of the node the item position range is passed of,
+// along with the node in which the item is present. This is done with the help of AST.
+// posRange can also be called as itemPosRange since carries the position range of the lexical item.
+func (n *nodeInfo) nodeHistory(head parser.Node, posRange parser.PositionRange, stack []parser.Node) ([]parser.Node, parser.Node) {
+	if head.PositionRange().Contains(posRange) {
+		stack = append(stack, head)
+	}
+	for _, child := range parser.Children(head) {
+		if child.PositionRange().Contains(posRange) {
+			return n.nodeHistory(child, posRange, stack)
+		}
+	}
+	return stack, head
+}
+
+// reduceNonNewLineExprs reduces those expressions from the history that are not
+// necessary for base indent. Base indent is applied on those items that have the
+// tendency to fall on a new line. However, items like Left_Bracket, Right_Bracket
+// never fall on new line. This means that when they (node in which they are present)
+// will be encountered, the stack already has MatrixSelector or SubQuery in it, but
+// their very existence harms the base indent. This is because they never contribute
+// to the indent, meaning the (metric_name[5m]) will get base indent as 3 (ParenExpr,
+// MatrixSelector, VectorSelector) but it should actually be 2 according to the requirements.
+// Hence, these unwanted expressions that do not contribute to the base indent
+// should be reduced.
+func reduceNonNewLineExprs(history []parser.Node) []parser.Node {
+	var temp []parser.Node
+	if !(len(history) > 1) {
+		return history
+	}
+	for i := range history {
+		if !(is(history[i], matrixExpr) || is(history[i], subQueryExpr)) {
+			temp = append(temp, history[i])
+		}
+	}
+	return reduceBinary(temp)
+}
+
+// reduceBinary reduces from end, the continuous occurrence
+// of binary expression to its single representation.
+func reduceBinary(history []parser.Node) []parser.Node {
+	var temp []parser.Node
+	if !(len(history) > 1) {
+		return history
+	}
+	for i := 0; i < len(history)-1; i++ {
+		if !(is(history[i], binaryExpr) && is(history[i+1], binaryExpr)) {
+			temp = append(temp, history[i])
+		}
+	}
+	temp = append(temp, history[len(history)-1])
+	return temp
+}
+
+func is(node parser.Node, typ uint) bool {
```

Comment:
`is` is a bit short of a function name, it looks like it's a Go thingy.

Reply:
updated to `compareNodeType`. How does that sound?

Reply:
Compare implies you're comparing two nodes. isNodeType should do it.

Reply:
Oh yes.

Reply:
Addressed this to `isNodeType`. Hence, resolving.